### PR TITLE
Prevent dangling cat-files                                           …

### LINKED ISSUE #17138: always running processes from 1.15.3
### Gitea Version

1.15.3

### Git Version

2.33.0.windows.2

### Operating System

windows

### How are you running Gitea?

running gitea by gitea-1.15.3-windows-4.0-amd64.exe

### Database

MySQL

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Description

always running processes from 1.15.3

the bug from 1.15.3

not appear in 1.15.2

### Screenshots

未命名1
未命名

### PR DESCRIPTION
**Pull request by zeripath**
In #17138 it appears that on Windows a git cat-file can fail to be killed.

This PR attempts to prevent this by:

1. Using os.Pipes for the input - this should mean that the stdin is definitely closed.
2. If the cat-file is doesn't close properly it will attempt to kill it repeatedly logging this.

Fix #17138

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
This is WIP because the current mechanism for killing processes is racy!

File: modules/git/batch_reader.go [hunk 1]
```diff
@@ -11,53 +11,70 @@ import (
...
-
 // CatFileBatchCheck opens git cat-file --batch-check in the provided repo and returns a stdin pipe, a stdout reader and cancel function
-func CatFileBatchCheck(repoPath string) (WriteCloserError, *bufio.Reader, func()) {
-	batchStdinReader, batchStdinWriter := io.Pipe()
+func CatFileBatchCheck(repoPath string) (io.WriteCloser, *bufio.Reader, func()) {
+	batchStdinReader, batchStdinWriter, err := os.Pipe()
+	if err != nil {
+		log.Critical("Unable to open pipe for cat-file --batch: %v", err)
+		rd, wr := io.Pipe()
+		rd.CloseWithError(err)
```
Comment (zeripath) [inline modules/git/batch_reader.go]:
```suggestion
		_ = rd.CloseWithError(err)
```

File: modules/git/batch_reader.go [hunk 2]
```diff
@@ -67,38 +84,60 @@ func CatFileBatchCheck(repoPath string) (WriteCloserError, *bufio.Reader, func()
...
-func CatFileBatch(repoPath string) (WriteCloserError, *bufio.Reader, func()) {
+func CatFileBatch(repoPath string) (io.WriteCloser, *bufio.Reader, func()) {
 	// We often want to feed the commits in order into cat-file --batch, followed by their trees and sub trees as necessary.
 	// so let's create a batch stdin and stdout
-	batchStdinReader, batchStdinWriter := io.Pipe()
+	batchStdinReader, batchStdinWriter, err := os.Pipe()
+	if err != nil {
+		log.Critical("Unable to open pipe for cat-file --batch: %v", err)
+		rd, wr := io.Pipe()
+		rd.CloseWithError(err)
```
Comment (zeripath) [inline modules/git/batch_reader.go]:
```suggestion
		_ = rd.CloseWithError(err)
```

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
data race detected

**Review by 6543 — APPROVED** [review]

**Comment by zeripath** [comment]
This doesn't work and the original issue appears to have solved itself.

**Pull request closed without merging**